File: 2156c.cpp
#include <bits/stdc++.h>

using namespace std;


int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;

    while (t--) {
        int n, k;
        cin >> n >> k;
        vector<int> cnt(n + 1, 0);
        for (int i = 0; i < n; i++) {
            int x;
            cin >> x;
            cnt[x]++;
        }
        vector<int>pref(n + 1, 0);
        for (int i = 1; i <= n; i++) pref[i] = pref[i - 1] + cnt[i];

        bool found = false;
        for (int i = n; i > 1; i--) {
            int s = pref[min(n, 4 * i - 1)] - cnt[i];
            if (2 * i <= n) s -= cnt[2 * i];
            if (3 * i <= n) s -= cnt[3 * i];
            if (s <= k) {
                cout << i << "\n";
                found = true;
                break;
            }
        }

        if (!found) cout << 1 << "\n";
    }

    return 0;
}

File: 2025/2156/2156c.py
t = int(input())
for _ in range(t):
    n, k = map(int, input().split())
    a = map(int, input().split())
    cnt = [0] * (n + 1)
    for x in a:
        cnt[x] += 1
    pref = [0] * (n + 1)
    for i in range(1, n + 1):
        pref[i] = pref[i - 1] + cnt[i]
    for i in range(n, 1, -1):
        s = pref[min(n, 4 * i - 1)] - cnt[i]
        if 2 * i <= n:
            s -= cnt[2 * i]
        if 3 * i <= n:
            s -= cnt[3 * i]
        if s <= k:
            print(i)
            break
    else:
        print(1)
